### PR TITLE
➕ Add FastFlow Model

### LINKED ISSUE #261: Add FastFlow Model
### Discussed in https://github.com/openvinotoolkit/anomalib/discussions/189

<div type='discussions-op-text'>

<sup>Originally posted by **innat** February 23, 2022</sup>
**Is your feature request related to a problem? Please describe.**

  - New sota: https://paperswithcode.com/sota/anomaly-detection-on-mvtec-ad
  - paper: https://arxiv.org/pdf/2111.07677v2.pdf
</div>

### PR DESCRIPTION
**Pull request by samet-akcay**
# Description

- Add fastflow implementation

- Fixes #261 

## Changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## Checklist

- [ ] My code follows the [pre-commit style and check guidelines](https://openvinotoolkit.github.io/anomalib/guides/using_pre_commit.html#pre-commit-hooks) of this project.
- [ ] I have performed a self-review of my code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing tests pass locally with my changes

### CODE REVIEW

File: anomalib/models/fastflow/__init__.py [hunk 1]
```diff
@@ -0,0 +1,10 @@
+"""FastFlow Algorithm Implementation."""
+
+# Copyright (C) 2022 Intel Corporation
+# SPDX-License-Identifier: Apache-2.0
+#
+
+from .lightning_model import Fastflow, FastflowLightning
+from .torch_model import FastflowLoss, FastflowModel
+
+__all__ = ["FastflowModel", "FastflowLoss", "FastflowLightning", "Fastflow"]
```
Comment (djdameln) [inline anomalib/models/fastflow/__init__.py:10]: Do we want to expose all these classes here in the init? For the other models we only list the `Lightning` class here. I'm fine with either option, but at least we should be consistent.

File: anomalib/models/fastflow/config.yaml [hunk 2]
```diff
@@ -0,0 +1,102 @@
+dataset:
+  name: mvtec #options: [mvtec, btech, folder]
+  format: mvtec
+  path: ./datasets/MVTec
+  task: segmentation
+  category: wood
+  image_size: 256 # options: [256, 256, 448, 384] - for each supported backbone
+  train_batch_size: 32
+  test_batch_size: 32
+  num_workers: 0
```
Comment (djdameln) [inline anomalib/models/fastflow/config.yaml:10]: `num_workers` is usually set to a higher default value, for the other models it's 36.

File: anomalib/models/fastflow/torch_model.py [hunk 3]
```diff
@@ -0,0 +1,250 @@
...
+                (hidden_variables, log-of-the-jacobian-determinants).
+                During the validation/test, return the anomaly map.
+        """
+        # pylint: disable=invalid-name
+
+        return_val: Union[Tuple[List[Tensor], List[Tensor]], Tensor]
+
+        self.feature_extractor.eval()
+        if isinstance(self.feature_extractor, VisionTransformer):
+            x = self.feature_extractor.patch_embed(x)
```
Comment (djdameln) [inline anomalib/models/fastflow/torch_model.py:198]: Does pylint allow single letter variable names?

File: anomalib/models/fastflow/torch_model.py [hunk 4]
```diff
@@ -0,0 +1,250 @@
...
+                create_fast_flow_block(
+                    input_dimensions=[channel, int(input_size[0] / scale), int(input_size[1] / scale)],
+                    conv3x3_only=conv3x3_only,
+                    hidden_ratio=hidden_ratio,
+                    flow_steps=flow_steps,
+                )
+            )
+        self.anomaly_map_generator = AnomalyMapGenerator(input_size=input_size)
+
+    def forward(self, x: Tensor) -> Union[Tuple[List[Tensor], List[Tensor]], Tensor]:
```
Comment (djdameln) [inline anomalib/models/fastflow/torch_model.py:181]: This method is a bit hard to read. Maybe you could split it into several sub-methods. Something like:
```
if isinstance(self.feature_extractor, VisionTransformer):
    features = self.get_transformer_features(x)
elif isinstance(self.feature_extractor, Cait):
    features = self.get_cait_features(x)
else:
    features = self.feature_extractor(x)
    
hidden_variables = self.get_hidden_variables(features)
```

**Review by djdameln — CHANGES_REQUESTED** [review]
Thanks! I have a few comments, mainly minor

File: anomalib/models/components/base/anomaly_module.py [hunk 5]
```diff
@@ -171,6 +171,8 @@ def _outputs_to_cpu(self, output):
 
     def _log_metrics(self):
         """Log computed performance metrics."""
-        self.log_dict(self.image_metrics)
         if self.pixel_metrics.update_called:
-            self.log_dict(self.pixel_metrics)
+            self.log_dict(self.pixel_metrics, prog_bar=True)
+            self.log_dict(self.image_metrics, prog_bar=False)
+        else:
+            self.log_dict(self.image_metrics, prog_bar=True)
```
[the remark this answers is not in the document]
Comment (samet-akcay) [reply anomalib/models/components/base/anomaly_module.py:178]: fastflow has long training time. without the performance being logged to the progress bar, it's hard to follow what's happening. 

pixel-metrics should be logged by default. However, if it's not available, image-metrics should be logged.

File: anomalib/models/fastflow/__init__.py (hunk 1 above)
Comment (samet-akcay) [reply anomalib/models/fastflow/__init__.py:10]: I think it would be good to be able to import them. Especially when we use them as components. We could do the same for the other algos

File: anomalib/models/fastflow/config.yaml (hunk 2 above)
Comment (samet-akcay) [reply anomalib/models/fastflow/config.yaml:10]: thanks for spotting this

File: anomalib/models/fastflow/torch_model.py (hunk 3 above)
Comment (samet-akcay) [reply anomalib/models/fastflow/torch_model.py:198]: I temporarily ignore that in L192. As I said, this `FastFlowModel` would require some refactor later.

**Review by ashwinvaidya17 — COMMENTED** [review]
Thanks! Only a few minor comments

File: anomalib/models/components/base/anomaly_module.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (ashwinvaidya17) [reply anomalib/models/components/base/anomaly_module.py:178]: I am clear. Maybe we can wait for @djdameln

**Review by ashwinvaidya17 — APPROVED** [review]

File: tests/nightly/models/test_model_nightly.py [hunk 6]
```diff
@@ -42,14 +42,15 @@ def get_model_nncf_cat() -> List:
         List: Returns a combination of models with their nncf support for each category.
     """
     model_support = [
-        ("padim", False),
+        ("cflow", False),
         ("dfkde", False),
         ("dfm", False),
-        ("stfpm", False),
+        ("fastflow", False),
```
Comment (ashwinvaidya17) [inline tests/nightly/models/test_model_nightly.py:48]: I don't think this will work as the performance thresholds for fastflow haven't been updated in the performance threshold file.

**Review by ashwinvaidya17 — CHANGES_REQUESTED** [review]

File: anomalib/models/fastflow/README.md [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+| Leather    |   1.000   |     0.999     | 0.999 | 0.991 |
+| Metal_nut  |   0.963   |     0.995     | 0.911 | 0.963 |
+| Pill       |   0.916   |     0.942     | 0.910 | 0.916 |
+| Screw      |   0.521   |     0.839     | 0.705 | 0.791 |
+| Tile       |   0.967   |     1.000     | 0.993 | 0.998 |
+| Toothbrush |   0.844   |     0.836     | 0.850 | 0.886 |
+| Transistor |   0.938   |     0.979     | 0.993 | 0.983 |
+| Wood       |   0.978   |     0.992     | 0.979 | 0.989 |
+| Zipper     |   0.878   |     0.951     | 0.981 | 0.977 |
+| Average    |           |               |       |       |
```
Comment (ashwinvaidya17) [inline anomalib/models/fastflow/README.md:51]: Can you add the average score? Also, a minor issue but this format is a bit different from our other models (rows <-> columns). Maybe for consistency we should decide on a single format. Maybe this can be addressed in a different PR.

Comment (samet-akcay) [reply anomalib/models/fastflow/README.md:51]: We could perhaps modify all the tables once we have all the benchmark results merged
